joomla/joomla-cms
# Allow Media Manager to show thumbnails for more file types.

### Summary of Changes

* Added new Media Manager “Preview Image Extensions (File Types)” configuration option.
* Updated `\Joomla\CMS\Helper\MediaHelper::isImage` to use the new option.
* Updated `\Joomla\CMS\Helper\MediaHelper::getMimeType` to return the correct MIME type even when EXIF returns the generic `application/octet-stream` type.
* Updated `\Joomla\Plugin\Filesystem\Local\Adapter\LocalAdapter::getPathInformation` to return a preview path even on image files without intrinsic dimensions (e.g. vector files such as SVG and EPS).

### Testing Instructions

* Install a new Joomla 4 site.
* Go to Content, Media, Options.
* Set “Check MIME Types” to No.
* Set “Legal Image Extensions (File Types)” to `bmp,gif,jpg,png,jpeg,webp,svg`
* Set “Preview Image Extensions (File Types)” to `bmp,gif,jpg,png,jpeg,webp,svg` (this step only applies AFTER applying the Pull Request)
* Upload the attached SVG file

### Actual result BEFORE applying this Pull Request

The SVG file is uploaded. Instead of a thumbnail of the file you see an eye icon with a slash through it.

### Expected result AFTER applying this Pull Request

The SVG file is uploaded. You see a thumbnail of the file.

### Further testing (with the PR applied)

**Test A**
* Go to Content, Media, Options.
* Set “Legal Image Extensions (File Types)” to `bmp,gif,jpg,png,jpeg,webp`
* Set “Preview Image Extensions (File Types)” to `bmp,gif,jpg,png,jpeg,webp,svg`

Expected result: you no longer see the uploaded SVG file at all. Moreover, you can no longer upload any SVG files. 

**Test B**
* Go to Content, Media, Options.
* Set “Legal Image Extensions (File Types)” to `bmp,gif,jpg,png,jpeg,webp,svg`
* Set “Preview Image Extensions (File Types)” to `bmp,gif,jpg,png,jpeg,webp`

Expected result: you see the SVG file, BUT instead of a thumbnail of the file you see an eye icon with a slash through it. You can upload other SVG files if you want.

### Documentation Changes Required

The "Preview Image Extensions (File Types)" option needs to be documented in https://help.joomla.org/proxy?keyref=Help4.x:Media:_Options/en as follows.

**Preview Image Extensions (File Types)**. Files whose extension is in this comma–separated list _and_ the “Legal Image Extensions (File Types)” will have previews (thumbnails) displayed for them in the Media Manager. If the file extension is in “Legal Image Extensions (File Types)” but not in the preview list the files will be displayed in the media manager but instead of a preview you will see an icon, either representing the file type or an eye with a slash going through it to indicate no preview and no file type icon is available. If the file extension is not in the “Legal Image Extensions (File Types)” — regardless of whether it is in the preview list — the file will NOT be shown at all in Media Manager.

### Translation impact

This Pull Request adds two new language strings in the front- _and_ backend language files of `com_media`:

* `COM_MEDIA_FIELD_PREVIEW_IMAGE_EXTENSIONS_LABEL`
* `COM_MEDIA_FIELD_PREVIEW_IMAGE_EXTENSIONS_DESC`

### Technical notes

#### Why not reuse the “Legal Image Extensions (File Types)” option?

Well, for two reasons. The “Legal Image Extensions (File Types)” controls which file types are allowed to be uploaded through the Media Manager **and** selected in Media fields. In many cases site integrators need to allow uploading and/or selecting PDF files which are all too common in the business and non–profit world. Whether treating PDF files are media files and/or using the Media Manager as a generic file manager is a good solution is highly debatable, but given the limited file management tools the core Joomla! CMS provides it's a reasonable solution.

So here's the first reason. PDF files tend to be big and do not have thumbnails. If we let PDFs to be previewed when they are merely allowed to be uploaded the Media Manager would get bogged down. Having 100 PDFs at 2.5MiB each on average would mean that accessing the Media Manager page for that folder would require transferring 250MiB of unnecessary data. Groan.

The second reason is that previewing and allowing files to be uploaded are two distinct use cases. For example, if I know my clients are on a rural area, thus severely bandwidth limited (2Mbps ADSL or even 56Kbps dial-up is still a thing in remote areas), I would much rather prefer them to NOT see any image previews at all.

#### Why change `getMimeType`?

The EXIF extension returns `application/octet-stream` for any file type it does not recognise. Since this extension is made to only work with _bitmap_ files that can carry EXIF data this makes sense. However, the way our code was written, if you had this extension installed on your site you could NEVER get the right MIME type for files not supported by the EXIF extension. Most notably, this meant we could never have SVG support.

The updated code uses fallbacks to try and get the real MIME type of the file more accurately.

#### Why change `\Joomla\Plugin\Filesystem\Local\Adapter\LocalAdapter::getPathInformation`?

There was a false assumption in the code that if something is an image it must _necessarily_ have intrinsic dimensions and that these dimensions can be read with `getimagesize`. The corollary was that image files without intrinsic dimensions could never get a preview image shown for them.

The first assumption is obviously wrong as vector images (e.g. SVG files) do not have intrinsic dimensions. That's the whole point of being vector graphics.

The second assumption is also wrong, since PHP only supports _at most_ the file types GIF, JPEG, JPEG200, PNG, SWF, PSD, BMP, WBMP, XBM, TIFF, IFF, JB2, JPC, JP2, JPX, SWC, ICO and WEBP. I say _at most_ because it depends on compile–time options. For instance, it's absolutely possible to have a PHP version which is not linked to `libpng`, thus lacks support for PNG files.

The way the code is rearranged decouples the ability to preview a file from its dimensions. After all, in Joomla 4 we are using a constant dimensions DIV with the file to be previewed as a background image, therefore we let the browser figure out the best way to present it while preserving the aspect ratio. Browsers are far better at this than PHP.

#### Why is the default setting for the new option identical to the legal image files?

To maintain the same behaviour as Joomla versions released to this day.

#### If I clear out the new option I can still see previews!

That's because an empty option is treated by Joomla! the same as having set it to the default (`bmp,gif,jpg,png,jpeg,webp`). Actually, if you empty this option and save you will see that upon the page reloading it's filled in with the default value. That's how Joomla has worked since at least version 1.6.0 released back in 2010. Don't blame me, I found it like that!

If you want to disable previews, set this option to a fake extension e.g. `invalid123`. Now all image files including PNG, JPG, WEBP etc will no longer display a preview.

_Maybe_ this could go into the documentation as it's non–obvious for most people?
